### PR TITLE
feat: Extract documentation from source files

### PR DESCRIPTION
**Pull request by gwhitney**
Adds a `semiliterate` configuration parameter to the `simple` plugin,
  consisting of a list of dicts describing how embedded documentation should
  be extracted from different types of source files. The default value for
  this parameter will extract markdown from `""" md` ... `"""` triple-quoted
  strings in Python files and `/** md` ... `**/` comment blocks in other
  source files. Full documentation in the package.

### CODE REVIEW

**Comment by gwhitney** [comment]
I think the release/deploy check, from me as a contributor, was supposed to fail. So I think all is well. If not, or if there is anything else I can do to enhance this PR, just let me know.

**Comment by gwhitney** [comment]
Oops, inadvertently left in a couple of debugging print()s, sorry.

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [inline mkdocs_simple_plugin/plugin.py:250]: Prefer to break this into two functions -- copy and extract

File: tests/integration/fixtures/ok-source-extract/module.grepout
Comment (athackst) [inline tests/integration/fixtures/ok-source-extract/module.grepout:3]: Shouldn't the "replace" option be able to remove leading whitespace?

I imagine this would be an issue in both C++ and python comments.

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:17]: By which you mean the README.md? Presuming so, I will move it there when I update the PR.

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:55]: Correct, that's why I'd like both the rescan and the to_extension parameters: so I can grab "user manual" and "api documentation" markdown from the same file, using two different extensions. Is it OK to leave in?

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:68]: I assume you mean that the 'begin' marker would serve as both start and resume, and that 'end' would serve as pause? I could switch to that, but this is a bit more flexible, in that: (A) by not specifying start but only pause and resume, you can indicate that extraction should start from the beginning of the file; and (B) you can have a marker that completely terminates (or even pre-empts) extraction from the file.

Since this was to be as language-agnostic as possible, I wanted as full flexibility in extracting snippets of code or documentation from arbitrary source files as possible, while keeping to fairly intuitive meanings, which I think start, pause, resume, end do. In my projects so far I have used start/pause, start/end, and pause/resume pairs productively.

Of course the PR could be adjusted. The easiest change would be to eliminate "resume," and merge its meaning with "start." As far as I can tell from my experience, the only real loss would be not having a way to start extraction from the first line of a file. I've never actually used a value of "resume" different from that for "start," which is why I put in the default for resume. (So I could also adjust the PR to have "start," "pause," and "end," plus a flag for whether the initial state is "extracting" or "not extracting" -- that's only 3 regexp parameters then and one flag, which is arguably simpler than 4 regexps. But I'd prefer to have both "pause" and "end" if you don't mind, their semantics are quite different, and end is useful because then you just don't have to worry about what the contents of the rest of the file are, you won't trigger unwanted extraction.)

File: mkdocs_simple_plugin/plugin.py
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.py:165]: Ah, I thought so, too. But unfortunately what I've found in my latest setup is that when I execute `mkdocs serve` it uses a tmp directory for site, so if there is a site directory lying around from a previous `mkdocs build` run, that site directory is *not* put in ignore_paths, and it gets searched for documentation, messing up the local server. So that's why I did this, but I would be happy for a cleverer solution if you have one.

File: mkdocs_simple_plugin/plugin.py
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.py:250]: OK, those two operations share a bunch of temporary local variable context which is why I lumped them, but I will just find the cleanest way to split them up in the revised PR.

File: tests/integration/fixtures/ok-source-extract/module.grepout
Comment (gwhitney) [reply tests/integration/fixtures/ok-source-extract/module.grepout:3]: Righto. The issue here is knowing how _much_ whitespace for "replace" to strip. Typically thatwould involve knowing how much whitespace there was at the beginning of the line of the "start" (and/or "resume", if we use it) marker. So I could save the start/resume match and allow backreferences to what it matched in the "replace" facility, but I didn't want to get too fancy on the first pass and I haven't actually used that facility yet. But I would be fine putting it in if you want me to -- let me know.

**Comment by gwhitney** [comment]
Responded to your code review comments -- thank you so much for the feedback! As soon as I get your thoughts/guidance on the "open" items of discussion I'll be happy to update the PR.

File: mkdocs_simple_plugin/plugin.md
Comment (athackst) [reply mkdocs_simple_plugin/plugin.md:55]: I think it might be cleaner to specify a "folder" instead of an extension to put the files in per pattern then? I imagine having an "API documentation" folder in the output could be pretty useful.

Also, why not always rescan?

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [inline mkdocs_simple_plugin/plugin.py:78]: Is 'pat' supposed to be short for 'pattern'?  Prefer whole words in function names.

File: mkdocs_simple_plugin/plugin.md
Comment (athackst) [reply mkdocs_simple_plugin/plugin.md:68]: Yes, I mean that it seems like "resume/pause" is the more useful parameters.

I'm not sure I understand the use case of wanting to grab from the beginning of a file and terminating halfway through?  What case are you trying to solve by 'end' that is not covered by 'pause'?

File: tests/integration/fixtures/ok-source-extract/module.grepout
Comment (athackst) [reply tests/integration/fixtures/ok-source-extract/module.grepout:3]: I wasn't thinking anything so fancy.  Markdown has limited use for indentation, so one could conceivably remove all whitespace from the beginning of the lines.  I'm ok with not including it and leave it to the user.

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:17]: Gotcha, will do.

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:55]: I was just using a to_extension value of "_api.md", so that "module.py" could produce both "module.md" (from the "regular" scan) and a "module_api.md". That seemed to work pretty well.

Being able to turn off rescan is helpful when you have mixed types of source files, say Python, C, and Java. You want a `(.py)$` pattern to grab """-strings, and want the fallback /** .. **/ pattern for both the C and Java files, so it has a permissive pattern `(.[^.])*`. You probably don't want to rescan the Python files for /** .. **/ -- it won't ever be useful, and in the worst case a trigger might inadvertently match some string in your Python code --  so you set rescan false on the fallback pattern.

So here are your options as I see them:

(1) On to_extension, I could 
  (A) change the parameter name to "destination" and have it be a template to be expanded by the filename match, which if not specified simply replaces the entire match with ".md". This would allow what I am currently doing, plus allow your quite reasonable proposed behavior of putting api rescans into a separate folder:
```
- pattern: '^(.*)\.py$'
  destination: 'api/\1.md'
```

  (2) I could leave to_extension as is, a string to replace the last matching group of the pattern;

  (3) Or I could eliminate to_extension altogether, which would seem to preclude double-extraction from the same file. If you think double-extraction is worthwhile for `simple` to support but you don't like either (1) or (2), I'm open to any alternate suggestion for how to support it.

(B) On rescan, I could
  (1) Leave it as is,
  (2) Remove this parameter and behave as if always "true" -- the user will simply have to ensure that at most one of the extraction triggers occurs in any file that matches multiple patterns.
  (3) Remove this parameter and behave as if always "false" -- like (A3) this would seem to preclude double extraction but I am open to other suggestions of how to support it if you would like to.

Let me know what you prefer.

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:68]: I was trying to provide user flexibility in the face of unknown languages/codebase arrangements that someone might be extracting documentation from.

For example, one could adopt a convention that there's Markdown in the opening comment block of (some kind of) scripts with `#`-marked line comments. Then one could write:
```
- pattern: '(.script)'      # whatever your script extension is
  replace: ['^#(.*\s*)$'] # strip leading `#`
  end: '^[^#]'                # terminate processing at end of beginning comment block
```
Then you can use comments normally in the rest of your script without worrying about accidentally creating more markdown. In general, since this proposed extraction facility for `simple` is operating via ad-hoc pattern matching, I see value in being able to terminate processing of a file altogether, so that below the end marker, totally free usage of whatever the file's primary language syntax is possible without worry of triggering additional markdown extraction.

Also, I was trying to create common parameters for markdown extraction and for the `{! ... !}` inclusion facility, which I so far have mainly used to extract source code examples into the documentation. When including, I generally have in mind "end" semantics rather than "pause" -- I am trying to extract one specific code snippet, and when I hit its end marker, I want extraction to stop. I don't want to inadvertently get more code because the start marker I chose happens to occur again later in the source file.

One other small example: in the semiliterate package, I defined 'end' for markdown files to be ''. That way, I could have a block at the end of README files explaining where to get further information, which would show in the repository view, but which is unnecessary and would be dropped in the documentation site. This is a fairly weak example, as 'pause' would work in practice there too, but the semantics I had in mind was "Hey, mkdocs, there's nothing else of interest to you in this file." The end parameter captures that, whereas pause at least suggests that there may be more content to come.

File: mkdocs_simple_plugin/plugin.py
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.py:165]: If by "bug" you mean "use a temp directory for site when doing mkdocs serve," that's standard behavior for mkdocs, see https://github.com/mkdocs/mkdocs/blob/master/mkdocs/commands/serve.py#L119. And as far as I know, standard behavior for "mkdocs build" is to leave the "site" dir behind in the project directory. Hence, you should find in your setup that "mkdocs build" followed by "mkdocs serve" is scanning the generated site for markdown files. Like this has simply gone under the radar since there generally aren't markdown files in the site directory. But you could force the issue by running "mkdocs build" and then hand-copying a markdown file into site/ and then running "mkdocs serve" and you should get a message about having a markdown file that's not mentioned in the navigation or something like that.

So anyhow, with more extensive extraction with this PR, it is definitely the case that it's more noticeable when a leftover site directory is being scanned under `mkdocs serve`. I do not think there is any "bug" involved. But should I file a separate issue for this point where we can discuss how you'd like to resolve it? I am happy to resolve it however you would like, but some action is needed for this PR to work with `mkdocs serve` in case there is a left-over site directory that should not be scanned. Thanks!

File: mkdocs_simple_plugin/plugin.py
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.py:78]: Gotcha, will rename when I revise the PR.

**Comment by gwhitney** [comment]
OK, have responded to latest round of feedback, thanks for the ongoing discussion. I guess I still need your guidance on where you'd like to end up with multiple-scanning parameters (where I originally proposed 'to_extension' and 'rescan') and with the extraction control parameters (all or some of 'start,pause,resume,end' or an alternate approach that you may prefer). You can make the call or happy to discuss more if you think it will be helpful. Finally, I suggested moving the discussion on the "scanning the site directory when `mkdocs serve` is executed" item to a separate issue because in fact it occurs independently of this PR, or I am happy to just resolve it here, whichever you prefer.

File: tests/integration/fixtures/ok-source-extract/module.grepout
Comment (gwhitney) [reply tests/integration/fixtures/ok-source-extract/module.grepout:3]: Fine; you can definitely remove all leading whitespace with the existing 'replace' facility, so I will leave this aspect be, at least for now.

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [reply mkdocs_simple_plugin/plugin.py:165]: The bug is that I expected the site_dir to be "site" within the folder and be ignored.  It looks like mkdocs serve is replacing it with a temp directory, which I wasn't aware of.

I think this looks like an ok solution for now.  I can think about the best way to get the original site_dir.

File: mkdocs_simple_plugin/plugin.md
Comment (athackst) [reply mkdocs_simple_plugin/plugin.md:68]: Ah, I see.  What you're describing is more akin to "front matter", whereas what I was expecting was more like a "code block".

I think it's fair to assume that people might comment their code in one of two styles
(1) The "front matter" style, where documentation is written at the top of the file.
(2) With "documentation blocks", where documentation is written throughout.

TL;DR I went back and forth on this since it seems like it should be reasonable to assume that the user can come up with a unique end string, but I'm ok with leaving a terminator.

From a philosophical point of view:

The main way in which people comment their code is via language standards.  Each one is unique and has a different syntax.  These "standard" versions of commenting code are well covered in other plugins (sphynx, mkdocstrings).  However, it would be more convenient to be able to write markdown documentation directly into source.  This is the purpose of this feature, it adds the ability to search through files for markdown snippets.  Therefore, it is reasonable to assume that users are going to be modifying the way they comment their code in order to use this feature.

Since users are already doing something against language standards to use this feature, I think it's reasonable to require a unique start/end tag.  I'm having a hard time coming up with examples in which makes finding a unique termination string difficult.  

The "front matter" case would mean that a user would only need to worry about the first instance of a string before terminating documentation, letting the first non-comment line signify the end of front matter documentation seems like a usable feature, even though it wouldn't be "unique".

File: mkdocs_simple_plugin/plugin.md
Comment (athackst) [reply mkdocs_simple_plugin/plugin.md:55]: (A)  
   (1) `pattern`/`destination` works for me.  `to_extension` makes me think you can export to`.txt` or `.pdf` or something, which it doesn't do.
(B) 
   (2) Pattern matches should occur on all files, all the time.  People should be prescriptive about what extensions they want to include for the pattern matching, and they could include the same type multiple times if they want to do some advanced configuration.

File: mkdocs_simple_plugin/plugin.md
Comment (gwhitney) [reply mkdocs_simple_plugin/plugin.md:68]: Thanks for putting so much thought into this! Here is where I _think_ you landed: extraction parameters "begin," "pause," and "end," with files always starting in "non-extracting" mode, both of what I called "start" and "resume" merged into "begin", and "pause" and "end" keeping the semantics I proposed. Then "documentation blocks" can be done by a combination of "begin" and "pause", and "front matter" can be done by a combination of setting "begin" to `^(.*\s*)$` and using "end".

I think that actually covers all of my current use cases, so if I have followed you correctly I am perfectly happy with this as a landing place. I will get started with all of the other items, and please confirm my understanding. Also, please weigh in if you have thoughts on the following point given where we've landed:

* Should "end" occurring before the first occurrence of "begin" pre-empt processing altogether? Right now it does, but in this setup I could go either way -- rather than having the occurrence of "end" pre-empt "begin," it could simply be the non-presence of "begin" that leads to nothing extracted from a file. That might make it easier when grabbing a code snippet as an example in documentation, to choose a known well-identified starting point (like a function name) and then close it at a possibly generic spot (like a bracket in the leftmost column) without being stymied because that end marker also occurs earlier in the file.

Thanks for considering.

**Comment by gwhitney** [comment]
Great! I now have plenty to get started, and on the extraction parameters I just wanted to confirm I understood where you ended up, and had one additional question that brought up (in the code review discussion about). Then I think that's everything.

File: mkdocs_simple_plugin/plugin.md
Comment (athackst) [reply mkdocs_simple_plugin/plugin.md:68]: > Thanks for putting so much thought into this! Here is where I think you landed: extraction parameters "begin," "pause," and "end," with files always starting in "non-extracting" mode, both of what I called "start" and "resume" merged into "begin", and "pause" and "end" keeping the semantics I proposed. Then "documentation blocks" can be done by a combination of "begin" and "pause", and "front matter" can be done by a combination of setting "begin" to ^(.*\s*)$ and using "end".

Yes, that sounds correct to me

>Should "end" occurring before the first occurrence of "begin" pre-empt processing altogether? Right now it does, but in this setup I could go either way -- rather than having the occurrence of "end" pre-empt "begin," it could simply be the non-presence of "begin" that leads to nothing extracted from a file. That might make it easier when grabbing a code snippet as an example in documentation, to choose a known well-identified starting point (like a function name) and then close it at a possibly generic spot (like a bracket in the leftmost column) without being stymied because that end marker also occurs earlier in the file.

I don't think it matters much.  If end == terminate extraction then nothing will get extracted, if begin doesn't occur then also nothing will get extracted, right?  I think both things can be true.  

I think you can grab your code snippet with "begin" and "pause", as there would be just one pair of those in the file (and your pause marker should be the first occurrence of that match after begin).

**Comment by gwhitney** [comment]
OK, that should implement everything according to what's hashed out above. Once again, thanks for the feedback.

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [inline mkdocs_simple_plugin/plugin.py:59]: style: prefer underscores for instance variables

```suggestion
        self.wrote_something = False
```

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [inline mkdocs_simple_plugin/plugin.py:68]: style: prefer full names for variables

```suggestion
    def check_pattern(self, pattern, line, emit_last=True):
```

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [inline mkdocs_simple_plugin/plugin.py:140]: FYI This shouldn't be needed once #87 lands

File: mkdocs_simple_plugin/plugin.md
Comment (athackst) [inline mkdocs_simple_plugin/plugin.md:70]: Slightly prefer "start"/"stop"/"end" instead of "begin"/"pause"/"end"

File: mkdocs_simple_plugin/plugin.py
Comment (athackst) [inline mkdocs_simple_plugin/plugin.py:258]: style: prefer underscores between words

```suggestion
        new_paths = []
```

**Review by athackst — COMMENTED** [review]
Some small style changes, but otherwise LGTM

Generally, I try to follow the google python style guide https://google.github.io/styleguide/pyguide.html#s3.16.2-naming-conventions

Thanks for this contribution!

**Review by athackst — APPROVED** [review]
LGTM

Thank you!

**Pull request merged**
